### PR TITLE
Add support for delivering of results of Jet queries to the client

### PR DESCRIPTION
**Pull request by viliam-durina**

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1201/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1201/Hazelcast-pr-builder-1201.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ClientDynamicConfigurationAwareConfigTest>DynamicConfigurationAwareConfigTest.testDecorateAllPublicMethodsFromTest:61 Method public com.hazelcast.config.SqlConfig com.hazelcast.config.Config.getSqlConfig() is declared by class com.hazelcast.config.Config whilst it should be declared by class com.hazelcast.client.impl.clientside.ClientDynamicClusterConfig
[ERROR]   DynamicConfigurationAwareConfigTest.testDecorateAllPublicMethodsFromTest:61 Method public com.hazelcast.config.SqlConfig com.hazelcast.config.Config.getSqlConfig() is declared by class com.hazelcast.config.Config whilst it should be declared by class com.hazelcast.internal.dynamicconfig.DynamicConfigurationAwareConfig
[ERROR] Errors: 
[ERROR]   DataSerializableImplementsVersionedTest.testDataSerializableForVersionedInterface ? StackOverflow
[INFO] 
[ERROR] Tests run: 31072, Failures: 2, Errors: 1, Skipped: 1005
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Comment by devozerov** [comment]
@viliam-durina Do you have an associated PR for Jet to see how it is integrated end-to-end? 
I do not understand why the client should distinguish between Jet and non-Jet results. All we return are rows, query ID, and possibly update count. There is nothing jet specific.

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1338/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1338/Hazelcast-pr-builder-1338.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   DataSerializableConventionsTest.test_dataSerializableClasses_areIdentifiedDataSerializable:121 There are 63 classes which are DataSerializable, not @BinaryInterface-annotated and are not IdentifiedDataSerializable.
[ERROR]   DataSerializableConventionsTest.test_serializableClasses_areIdentifiedDataSerializable:161 There are 56 classes which are Serializable, not @BinaryInterface-annotated and are not IdentifiedDataSerializable.
[INFO] 
[ERROR] Tests run: 3857, Failures: 2, Errors: 0, Skipped: 14
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

File: hazelcast/src/main/java/com/hazelcast/sql/impl/client/QueryClientStateRegistry.java [hunk 1]
```diff
@@ -100,14 +102,16 @@ private boolean fetchPage(
...
         while (iterator.hasNext()) {
             SqlRow row = iterator.next();
             Row rowInternal = ((SqlRowImpl) row).getDelegate();
             Data rowData = serializationService.toData(rowInternal);
 
             page.add(rowData);
 
-            if (page.size() == cursorBufferSize) {
+            // TODO we call nanoTime for each item - use batching
+            if (page.size() == cursorBufferSize || System.nanoTime() >= endTime) {
```
Comment (devozerov) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/client/QueryClientStateRegistry.java:114]: 1) Isn't System.currentTimeMillis enough here? (though, I do not know any performance difference between `nano` and `millis`
2) The last line in this method calls `hasNext` which is blocking as well. Therefore, the idea of early return doesn't really work in this prototype. Instead, the method should return as soon as the very first call to `hasNext` returned something

File: hazelcast/src/main/java/com/hazelcast/sql/impl/exec/root/BlockingRootResultConsumer.java [hunk 2]
```diff
@@ -35,7 +34,7 @@
     private final InternalIterator iterator = new InternalIterator();
 
     /** Query context to schedule root execution when the next batch is needed. */
-    private volatile QueryFragmentContext context;
+    private volatile Runnable scheduleCallback;
```
Comment (devozerov) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/exec/root/BlockingRootResultConsumer.java:37]: Let's avoid `Runnable` and introduce a separate interface with proper name.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/client/QueryClientStateRegistry.java (hunk 1 above)
Comment (viliam-durina) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/client/QueryClientStateRegistry.java:114]: `nanoTime()` should always be used for elapsed time. CTM isn't monotonic in rare cases, doesn't run at constant speed. Performance differs by OS and JVM implementation, but I think on Linux they are comparable. IIRC on Windows CTM is faster, but also much less precise (~15ms resolution)

Comment (viliam-durina) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/client/QueryClientStateRegistry.java:114]: problem #2: fixed

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1357/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1357/Hazelcast-pr-builder-1357.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/sql/impl/client/QueryClientStateRegistry.java:110:24: Inner assignments should be avoided. [InnerAssignment]
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1358/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1358/Hazelcast-pr-builder-1358.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   DataSerializableConventionsTest.test_dataSerializableClasses_areIdentifiedDataSerializable:121 There are 63 classes which are DataSerializable, not @BinaryInterface-annotated and are not IdentifiedDataSerializable.
[ERROR]   DataSerializableConventionsTest.test_serializableClasses_areIdentifiedDataSerializable:161 There are 56 classes which are Serializable, not @BinaryInterface-annotated and are not IdentifiedDataSerializable.
[INFO] 
[ERROR] Tests run: 3857, Failures: 2, Errors: 0, Skipped: 14
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Pull request merged**